### PR TITLE
gnucash: Fix compilation w/ newer glib/gcc

### PR DESCRIPTION
**Pull request by rski**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 220751` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gnucash</li>
  </ul>
</details>

**Review by GabrielDougherty — APPROVED** [review]

**Comment by jtojnar** [comment]
Does not appear to be enough

```
In file included from /nix/store/xa7cna4ymrq7p67ykx1ddjlsigpqm0xp-glib-2.76.0-dev/include/glib-2.0/glib/giochannel.h:36,
                 from /nix/store/xa7cna4ymrq7p67ykx1ddjlsigpqm0xp-glib-2.76.0-dev/include/glib-2.0/glib.h:56,
                 from /build/gnucash-4.12/libgnucash/app-utils/file-utils.c:24:
/build/gnucash-4.12/libgnucash/app-utils/file-utils.c: In function 'gnc_getline':
/nix/store/xa7cna4ymrq7p67ykx1ddjlsigpqm0xp-glib-2.76.0-dev/include/glib-2.0/glib/gstring.h:72:5: error: ignoring return value of 'g_string_free_and_steal' declared with attribute 'warn_unused_result' [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result-Werror=unused-result8;;]
   68 |   (__builtin_constant_p (free_segment) ?        \
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   69 |     ((free_segment) ?                           \
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   70 |       (g_string_free) ((str), (free_segment)) : \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   71 |       g_string_free_and_steal (str))            \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   72 |     :                                           \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   73 |     (g_string_free) ((str), (free_segment)))
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/gnucash-4.12/libgnucash/app-utils/file-utils.c:149:5: note: in expansion of macro 'g_string_free'
  149 |     g_string_free(gs, FALSE);
      |     ^~~~~~~~~~~~~
```

**Review by AndersonTorres — APPROVED** [review]
My only complain is a bit pedantic: give names to the patches. Noneteheless, approve.

**Comment by bobby285271** [comment]
I can confirm this PR \*\*is\*\* not enough as jtojnar said, you need to cherry-pick this commit to the gnome branch before build since glib 2.76 is \*\*not\*\* on master branch and ofborg/nixpkgs-review is not useful if the merge base is master.

**Comment by bobby285271** [comment]
To fix building with both gcc 13 and glib 2.76, a quick search shows me four patches:

- https://github.com/Gnucash/gnucash/commit/ce3447e6ea8b2f734b24a2502e865ebbbc21aaaa
- https://github.com/Gnucash/gnucash/commit/184669f517744ac7be6e420e5e1f359384f676d5
- https://github.com/Gnucash/gnucash/commit/abcce5000ca72bf943ca8951867729942388848e
- https://github.com/Gnucash/gnucash/commit/89e63ef67235d231d242f018894295a6cb38cfc3

([What does `bis` and `ter` in the commit message mean?](https://stackoverflow.com/questions/69235095/suffix-bis-ter-in-bulma-what-do-they-mean))

For some reason this still doesn't build for me as the [`test_xaccFreeAccount`](https://github.com/Gnucash/gnucash/blob/7a4a959a9e316fe03431c9ea2333cf437ce74baa/libgnucash/engine/test/utest-Account.cpp#L859) test fails :confused: But I am testing this against 52ada08dc8f91afdd4fb44d530643cb211301e89 (i.e. glib 2.75.4) since there is not webkitgtk cache on current gnome branch this may or may not be a false positive, I may re-try later™

<details><summary>Some not-so-useful build logs</summary>

```
 50/115 Test  #34: test-engine ......................Bus error***Exception:   0.18 sec
TAP version 13
# random seed: R02Sd803b1dfd432d4c55cf437d33e3baab9
1..179
# Start of engine tests
# Start of Account tests
ok 1 /engine/Account/gnc set account separator
ok 2 /engine/Account/gnc account name violations errmsg
ok 3 /engine/Account/gnc account list name violations
ok 4 /engine/Account/account create and destroy
ok 5 /engine/Account/xaccMallocAccount
ok 6 /engine/Account/gnc account create root
ok 7 /engine/Account/xaccCloneAccount
ok 8 /engine/Account/xaccFreeAccountChildren

        Start  58: test-qofquerycore
 51/115 Test  #53: test-qofsession ..................   Passed    0.02 sec
```

</details>

**Comment by rski** [comment]
> mentions glib**c**

whoops, force of habit

**Comment by rski** [comment]
I've updated the list of patches and fix the glibc mention

**Comment by bobby285271** [comment]
> For some reason after fetching these patches this still doesn't build for me as the [test_xaccFreeAccount](https://github.com/Gnucash/gnucash/blob/7a4a959a9e316fe03431c9ea2333cf437ce74baa/libgnucash/engine/test/utest-Account.cpp#L859) test fails

Actually there is a fix for test-engine

- https://github.com/Gnucash/gnucash/commit/1020bde89c77f70cee6cc8181ead96e8fade47aa

File: pkgs/applications/office/gnucash/default.nix [hunk 1]
```diff
@@ -74,6 +75,28 @@ stdenv.mkDerivation rec {
     ./0003-remove-valgrind.patch
     # this patch makes gnucash exec the Finance::Quote helpers directly
     ./0004-exec-fq-helpers.patch
+    # the next two patches fix compilation with gcc 13 and glib > 2.76
```
Comment (bobby285271) [inline pkgs/applications/office/gnucash/default.nix:78]: Also no longer *two* patches :-)

**Comment by rski** [comment]
> > For some reason after fetching these patches this still doesn't build for me as the [test_xaccFreeAccount](https://github.com/Gnucash/gnucash/blob/7a4a959a9e316fe03431c9ea2333cf437ce74baa/libgnucash/engine/test/utest-Account.cpp#L859) test fails
> 
> Actually there is a fix for this, ~I will test this in 12 hours~
> 
>     * [Gnucash/gnucash@1020bde](https://github.com/Gnucash/gnucash/commit/1020bde89c77f70cee6cc8181ead96e8fade47aa)
> 
> 
> This builds for me after picking this

nice sleuthing. I've included this patch as well

File: pkgs/applications/office/gnucash/default.nix [hunk 2]
```diff
@@ -74,6 +75,34 @@ stdenv.mkDerivation rec {
...
+    # "Protect against passing an lseek failure rv to read()."
+    (fetchpatch2 {
+      url = "https://github.com/Gnucash/gnucash/commit/ce3447e6ea8b2f734b24a2502e865ebbbc21aaaa.patch";
+      sha256 = "sha256-mfPs/5rkCamihE0z1SRoX0tV4FNPkKUGd1T6iaCwy7E=";
+    })
+    # "Fix crashes in test-engine on Arch Linux."
+    # Also fixes the same crashes in nixpkgs.
+    (fetchpatch2 {
+      url = "https://github.com/Gnucash/gnucash/commit/1020bde89c77f70cee6cc8181ead96e8fade47aa.patch";
+      sha256 = "sha256-JCWm3M8hdgAwjuhLbFRN4Vk3BQqpn0FUwHk6Kg5Qa7Q=";
```
Comment (AndersonTorres) [inline pkgs/applications/office/gnucash/default.nix:104]: `fetchpatch2` does not support `hash`?

Comment (rski) [reply pkgs/applications/office/gnucash/default.nix:104]: guess it does, and anything but `hash` seems to be the old way of specifying the hash in fetchurl.  I'll clean it up when i i find some time

**Pull request merged**